enable integer multiplication/divison on FiniteDuration, see SI-6389

rkuhn opened the pull request:
This fixes a regression introduced in my previous pull request (despite all the review), which made duration arithmetic limited to addition / subtraction in practice (due to losing the FiniteDuration static type).

Review by: @phaller @axel22 @jsuereth

rkuhn committed:
enable integer multiplication/divison on FiniteDuration, see SI-6389

- without this "2.seconds * 2" will not return FiniteDuration but
  Duration
- added test case verifying this behavior
- matches normal arithmetics: integers throw on div-by-zero while
  doubles yield infinity; extended here to include overflow protection
  on integer multiplication

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/494/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1202/

jsuereth commented on src/library/scala/concurrent/util/Duration.scala:
```diff
@@ -704,16 +704,47 @@ final class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duratio
+   * @throws IllegalArgumentException if the result would overflow the range of FiniteDuration
+   */
+  def *(factor: Long) = {
+    if (length > Long.MaxValue / factor) throw new IllegalArgumentException("multiplication overflow")
```
This seems an expensive check.  Is there a _safe_ way to optimistically do the calculation, then detect overflow?  If not, LGTM.

rkuhn replied:
I’ll check (and benchmark).

rkuhn replied:
I have some initial thoughts about how to do it using Long.numberOfLeadingZeros, benchmarking suggests that that would bring the check down from 10ns to 0.3ns, but the code would look highly non-obvious as a result, so I don’t think it would be worth it.

rkuhn replied:
don't merge, even if otherwise tempted, I just saw a bug on the line we are currently discussing

soc replied:
How can a `Long` be possibly larger than `Long.MaxValue`? :-) EDIT: Sorry missed it completely ...

Additionally: `def /(factor: Long)` ... shouldn't the parameter be called `divisor`?

axel22 replied:
Given that `length` is an integer number (i.e. it's not real), `length > Long.MaxValue / factor` should logically (i.e. ignoring overflows) be the same as `length * factor > Long.MaxValue`.
So - this check looks ok to me.

As for performance, possibly counting the number of leading zeros in `factor` and `length`, and only then reverting to the more expensive check, could be faster. I suppose that's how you were doing it, Roland?

rkuhn replied:
actually I think I have found a solution which is reasonably comprehensible and does not use division, I’m currently fighting ScalaCheck in order to get it approved ;-)

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/494/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1202/

rkuhn committed:
improve performance of integer multiplication overflow check on Duration

- also rename divisor arguments to “divisor”
- and add a scalacheck for multiplication overflow detection

rkuhn committed:
move Duration (incl. DSL) into scala.concurrent.duration package

so that the full package can be imported naturally:

import scala.concurrent.duration._

will give you all the types (Duration, FiniteDuration, Deadline) and the
DSL for constructing these.

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/517/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/517/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1226/

paulp committed:
Adjustments to scala.concurrent.duration.

More use of implicit classes and value classes; aliased units to
make importing TimeUnit and TimeUnit._ unnecessary; placed some
classes in their own files because "the unit of compilation is
the file" and we shouldn't bundle more than necessary; fixed some
examples.

paulp commented:
I have a followup: #1355 .

axel22 commented:
LGTM

rkuhn committed:
some small remaining fixes

- added test for “span” and “fromNow” qualifiers
- make those actually work even when there is an expected type
- add ScalaDoc to them
- verify (and fix) conversion Deadline -> FiniteDuration
- also make Int * Duration => FiniteDuration work (and test it)

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/533/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1241/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/533/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1241/

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/536/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1244/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/536/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1244/

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/542/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1250/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1250/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/542/

rkuhn commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/546/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1254/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/546/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1254/

paulp commented:
It looks like a winner to me.
